book/scheduling-and-threading.md

```diff
@@ -98,6 +102,167 @@ class TaskQueue:
+without having to dive into the guts of a `Tab` or how it loads web pages---all
+we'd have to do is implement a new `TaskRunner` heuristic.
+
+Now for the fun part I promised. Let's implement the
+[`setTimeout`][settimeout] JavaScript API, which provides a way to run
+JavaScript a given number of milliseconds from now. In terms of the JavaScript
+and Python communication, it'll use an approach with handles, similar to the
+`addEventListener` code we added in [Chapter 9](scripts.md#event-handling).
+The *new* part will be our first use of Python threads.
+
+[settimeout]: https://developer.mozilla.org/en-US/docs/Web/API/setTimeout
+
+We'll very soon have a lot more use for threads; for now all you need to know is
+that the `threading` module allows you to create threads and run code on them
+now or at a time specified in the future. To use them, start by adding an import
+for that module:
+
+``` {.python}
+import threading
+```
+
+Implement a `set_timeout` helper function that runs a callback at a specified
+time in the future. You can do that by creating a new
+[Python thread][python-thread] via the `threading.Timer` class, which takes
+two parameters: a time delta in seconds from now, and a function to call when
+that time expires.
+
+[python-thread]: https://docs.python.org/3/library/threading.html
+
+``` {.python}
+def set_timeout(func, sec):     
+    t = threading.Timer(sec, func)
+    t.start()
+```
+
+Now we're ready to implement `setTimeout`. In the JavaScript runtime, add
+a new internal handle for each call to `setTimeout`, and store the mapping
+between handles and callback functions in a global object called
+`SET_TIMEOUT_REQUESTS`. When the timeout occurs, Python will call
+`__runSetTimeout` and be passed the handle.
+
+``` {.javascript file=runtime}
+SET_TIMEOUT_REQUESTS = {}
+
+function setTimeout(callback, time_delta) {
+    var handle = Object.keys(SET_TIMEOUT_REQUESTS).length;
+    SET_TIMEOUT_REQUESTS[handle] = callback;
+    call_python("setTimeout", handle, time_delta)
+}
+
+function __runSetTimeout(handle) {
+    var callback = SET_TIMEOUT_REQUESTS[handle]
+    delete SET_TIMEOUT_REQUESTS[handle];
+    callback();
+}
+```
+
+On the Python side, add a binding for `setTimeout` and an implementation that
+calls `set_timeout`. However, we have to be careful here, since in the code
+below, `run_callback` will run *on a different Python thread than the current
+one*. So we can't just call `evaljs` directly, or we'll end up with JavaScript
+running on two Python threads at the same time, which is not ok.[^js-thread]
+
+Not all is lost though! This problem can be solved elegantly with the
+`TaskRunner`: instead of running the script right away, schedule a task to
+do it later, when the other thread is free.[^locking] Here's the code:
+
+[^locking]: This code has a bug: it accesses data structures shared
+between two threads without a thread lock. We'll see soon how to fix it.
```

Comment:
ATM this chapter moves to async rendering first, and introduces all the necessary locks as it goes. I'm wondering whether, with `setTimeout`, there's a chance to introduce all the locks and discuss cross-thread communication first, before splitting off rendering. That would hopefully be a bit simpler?

Reply:
Done in PR 340.